packages/k8s.contrib.crd/internal/ModuleGenerator.pkl

```diff
@@ -0,0 +1,366 @@
+  if (schema is Boolean)
+    null
+  else
+    let (docCommentText: String =
+      List(
+        schema.title,
+        schema.description,
+        if (type is "module") "This module was generated from the CustomResourceDefinition at <\(baseUri)>." else null,
+        if (type is "class"|"module") getWarnings(schema, type) else null,
+        if (schema.default != null) "Default if undefined: `\(pcfRenderer.renderValue(schema.default))`" else null
+      )
+        .filterNonNull()
+        .join("\n\n")
+    )
+      if (docCommentText.isEmpty) null
+      else new DocCommentNode {
+        value = docCommentText
+        autoWrap = true
+      }
+
+local function getWarnings(schema: JsonSchema, type: "class"|"module"): String? =
+  if (type == "class") null
+  else if (schema.type == "object")
+    "WARN: The root schema describes open-ended properties, but this is not possible to describe at the module level."
+  else if (schema.type != null)
+    "WARN: The root schema's type is `\(jsonRenderer.renderValue(schema.type))`, and cannot be correctly mapped to a Pkl module."
+  else
+    "WARN: unable to determine module-level properties on the root schema."
+
+local function generateClassBody(
+  schema: JsonSchema(this.properties != null),
+  typeNames: Type.TypeNames
+): Listing<ClassOrModuleNode.PropertyDefinitionNode> =
+  new {
+    for (propName, propSchema in schema.properties!!) {
+      new {
+        name {
+          value = propName
+        }
+        when (propSchema is JsonSchema && propSchema.deprecated == true) {
+          annotations { utils.DEPRECATED }
+        }
+        docComment = getDocComment(propSchema, null)
+        typeAnnotation {
+          // If this property doesn't appear in the `required` array, mark it as nullable.
+          // We can't do this within [TypesGenerator] because it doesn't have enough information available.
+          type = let (underlyingType = typesGenerator.generateTypeNode(propSchema, typeNames))
+              if (schema.required?.toSet()?.contains(propName) ?? false) underlyingType
+              // If the type is already nullable, no need to make it *more* nullable.
+              else if (underlyingType is TypeNode.NullableTypeNode) underlyingType
+              else new TypeNode.NullableTypeNode { typeNode = underlyingType }
+        }
+        defaultValue =
+          if (typeAnnotation.type is TypeNode.UnionTypeNode
+          && (typeAnnotation.type as TypeNode.UnionTypeNode).members[0] is TypeNode.StringLiteralTypeNode
+          )
+            // If the first member is a string literal type, explicitly set the default to Undefined().
+            // We need to do this because the implicit default of a union type is the default of the first member.
+            // However, this is an incorrect way to represent enums in JSON Schema, because there should be no
+            // bias towards any of the values.
```

Comment:
This is not true in Pkl. It has a `*` annotation to tell you which member of a type union should provide the default. Unions without a `*` have no default value.